homeassistant/components/rhasspy/stt.py

```diff
@@ -0,0 +1,197 @@
+    """Set up Rhasspy speech to text component."""
+    provider = RhasspySTTProvider(hass, config)
+    _LOGGER.debug("Loaded Rhasspy stt provider")
+
+    # Register WAV speech to text test endpoint
+    hass.http.register_view(RhasspyWavView(hass))
+
+    return provider
+
+
+def get_speech_url(hass):
+    """Ties to get API URL from Rhasspy provider. Falls back to default."""
+    provider = hass.data.get(DOMAIN)
+    if provider is not None:
+        # Use provider URL
+        return urljoin(provider.api_url, "speech-to-text")
+
+    # Use default
+    return DEFAULT_SPEECH_URL
+
+
+# -----------------------------------------------------------------------------
+
+
+class RhasspySTTProvider(Provider):
+    """Rhasspy speech to text provider."""
+
+    def __init__(self, hass, conf):
+        """Create Rhasspy speech to text provider."""
+        self.hass = hass
+        self.config = conf
+
+        # URL to stream microphone audio
+        self.speech_url = conf.get(CONF_SPEECH_URL, None)
+        self.headers = {"Content-Type": "audio/wav"}
+
+    async def async_process_audio_stream(
+        self, metadata: SpeechMetadata, stream: aiohttp.StreamReader
+    ) -> SpeechResult:
+        """Process an audio stream to STT service.
+
+        Only streaming of content are allow!
+        """
+        if self.speech_url is None:
+            self.speech_url = get_speech_url(self.hass)
+
+        _LOGGER.debug("Receiving audio")
+        text_result = ""
+
+        try:
+            # First chunk is a WAV header (no frames)
+            header_chunk = True
+            with io.BytesIO() as wav_io:
+                wav_file = wave.open(wav_io, "wb")
+                async for audio_chunk, _ in stream.iter_chunks():
+                    if header_chunk:
+                        # Extract WAV information
+                        header_chunk = False
+                        with io.BytesIO(audio_chunk) as header_io:
+                            with wave.open(header_io) as header_file:
+                                wav_file.setnchannels(header_file.getnchannels())
+                                wav_file.setsampwidth(header_file.getsampwidth())
+                                wav_file.setframerate(header_file.getframerate())
+                    else:
+                        # Everything after first chunk is audio data.
+                        # Add to in-memory WAV file.
+                        wav_file.writeframes(audio_chunk)
+
+                wav_file.close()
+                wav_data = wav_io.getvalue()
+                _LOGGER.debug("Received %s byte(s)", len(wav_data))
+
+            # POST to Rhasspy server
+            session = async_get_clientsession(self.hass)
+            async with session.post(
+                self.speech_url, headers=self.headers, data=wav_data
+            ) as res:
+                res.raise_for_status()
+                text_result = await res.text()
+                _LOGGER.debug(text_result)
```

Comment:
You need write this into a library: https://developers.home-assistant.io/docs/en/creating_platform_code_review.html#5-communication-with-devices-services

Reply:
OK, in progress.

Reply:
Now using [rhasspy-client](https://pypi.org/project/rhasspy-client/)